[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression

wayi1 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#50283 [Gradient Compression] Fix warm-start for PowerSGD laywerwise compression**

Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Also fix the unit test in distributed_test.py. Previously the process group was not created correctly, and not communication occurred in the test_DistributedDataParallel_powerSGD_ddp_comm_hook.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

Commit:
[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression

Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression"


Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]

rohan-varma commented on torch/distributed/algorithms/ddp_comm_hooks/powerSGD_hook.py:
```diff
@@ -232,21 +234,26 @@ def powerSGD_hook(state: PowerSGDState, bucket) -> torch.futures.Future:
+                    device="cpu",
+                    dtype=dtype,
+                ).to(device)
+                _orthogonalize(q)
```
Nit: we call `_orthogonalize(q)` in both branches, will it be clearer to do:
```
for q in qs:
    _orthog(q) 
```

only once after the if/else block?

SciPioneer replied:
> Nit: we call `_orthogonalize(q)` in both branches, will it be clearer to do:
> 
> ```
> for q in qs:
>     _orthog(q) 
> ```
> 
> only once after the if/else block?

Agree that calling `_orthogonalize(q)` in a loop in just one branch can make the code more clear. However, this means two loops if warm-start is disabled: 1) create matrices; 2) orthogonalize these matrices.


In contrast, the current implementation can leverage the data locality, and hence it's better for the performance -- once q is created with random values, orthogonalize q immediately when it's still hot (i.e., in cache).

rohan-varma commented:
Also, looks like a related UT `test_DistributedDataParallel_powerSGD_ddp_comm_hook` is failing with this change.

SciPioneer commented:
> Also, looks like a related UT `test_DistributedDataParallel_powerSGD_ddp_comm_hook` is failing with this change.

Yeah, did notice that. Tried to debug this test, but the values after reduce somehow do not seem to be summed up. Will do more debugging later this week.

Commit:
Update on "[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression"


Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Also fix the unit test in distributed_test.py. Previously the process group was not created correctly, and not communication occurred in the test_DistributedDataParallel_powerSGD_ddp_comm_hook.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression"


Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Also fix the unit test in distributed_test.py. Previously the process group was not created correctly, and not communication occurred in the test_DistributedDataParallel_powerSGD_ddp_comm_hook.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression"


Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Also fix the unit test in distributed_test.py. Previously the process group was not created correctly, and not communication occurred in the test_DistributedDataParallel_powerSGD_ddp_comm_hook.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]

Commit:
Update on "[Gradient Compression] Fix warm-start for PowerSGD laywerwise compression"


Realize that for the layerwise compression, the previous warm-start implementation only skips memory allocations, but does not skip filling random values for Qs.

Also fix the unit test in distributed_test.py. Previously the process group was not created correctly, and not communication occurred in the test_DistributedDataParallel_powerSGD_ddp_comm_hook.

Original PR issue: Investigate Applying PowerSGD to Communication Hook for Gradient Compression #47202

Differential Revision: [D25846222](https://our.internmc.facebook.com/intern/diff/D25846222/)

[ghstack-poisoned]
